client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -122,10 +146,16 @@ private void handleMissingLog() {
      * message find which new segment it should go to and send it there. This can happen recursively if segments turn
      * over very quickly.
      */
-    @GuardedBy("lock")
     private void handleLogSealed(Segment segment) {
-        List<PendingEvent> toResend = selector.refreshSegmentEventWritersUponSealed(segment);
-        resend(toResend);
+        //TODO: shrids since this is executed by the callback for that segment we need to pass the events to be resent.
+        lock.readLock().lock();
+        try {
+            List<PendingEvent> toResend = selector.refreshSegmentEventWritersUponSealed(segment,
+                    segmentSealedCallBack);
+            resend(toResend);
```

Comment:
If during resend one of the segments being written to is sealed it will invoke this method which will then result in the two re-send calls going in parallel (As this is only protected by a read lock).  Previously the semantic was that resend would succeed or it would add it back to the `unsent` list and loop on it again. This has the effect of preserving the order of the events. To maintain this we might need a `trySend` method that returns a boolean or something.